src/encodings.cxx

```diff
@@ -637,8 +637,8 @@ encoding_group enc_group(std::string_view encoding_name)
 
   auto const found_encoding_group{encoding_map.find(encoding_name)};
   if (found_encoding_group == encoding_map.end())
-    throw std::invalid_argument{"unrecognized encoding '" +
-                                std::string{encoding_name} + "'"};
+    throw std::invalid_argument{"Unrecognized encoding: '" +
+                                std::string{encoding_name} + "'."};
```

Comment:
Inefficient string concatenation.

"unrecognized encoding '" gets converted to an std::string, causing an allocate-and-copy
std::string{encoding_name} gets converted to an std::string, causing an allocate-and-copy
"'." gets converted to an std::string, causing an allocate-and-copy

Then "unrecognized encoding '" + std::string{encoding_name} construct a new string object, which causes a new allocate-and-copy,

and finally the result of the previous string gets operator+'ed with "'.", which results in a final allocate and copy.

You want something like this instead: https://github.com/splinter-build/splinter/blob/10-string_concat/src/string_concat.h

Which does not require any intermediate std::strings at all, but instead calculates the total length of the final string, and allocates a single string object with that amount of space, then copies all data into it one time.

Reply:
Indeed.  I didn't bother with that here because it should be a relatively rare failure path, but you'd expect something like your string concatenation to end up in the standard library eventually.

Reply:
Something like this, I imagine: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1228r1.html